Sources/_StringProcessing/MatchingOptions.swift

```diff
@@ -0,0 +1,151 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2021-2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import _MatchingEngine
+
+/// A type that represents the current state of regex matching options, with
+/// stack-based scoping.
+struct MatchingOptions {
+  /// A set of matching options. 
+  fileprivate struct Representation: OptionSet, RawRepresentable {
+    var rawValue: UInt32
+
+    /// Options that comprise the mutually exclusive test segmentation group.
+    static var textSegmentOptions: Self {
+      Self(unchecked: .textSegmentGraphemeMode, .textSegmentWordMode)
+    }
+
+    /// Options that comprise the mutually exclusive semantic matching level
+    /// group.
+    static var semanticMatchingLevels: Self {
+      Self(unchecked: .graphemeClusterSemantics, .unicodeScalarSemantics, .byteSemantics)
+    }
+
+    /// The default set of options.
+    static var `default`: Self {
+      Self(unchecked: .graphemeClusterSemantics, .textSegmentGraphemeMode)
+    }
+
+    func contains(_ kind: AST.MatchingOption.Kind) -> Bool {
+      self.rawValue & (1 << kind.rawValue) != 0
+    }
+    
+    /// Merges `sequence` with this option set, preserving the
+    mutating func merge(with sequence: AST.MatchingOptionSequence) {
+      if sequence.caretLoc != nil {
+        self = .default
+      }
+      
+      for opt in sequence.adding {
+        // If opt is in one of the mutually exclusive groups, clear out the
+        // group before inserting.
+        if opt.isSemanticMatchingLevel {
+          remove(.semanticMatchingLevels)
+        }
+        if opt.isTextSegmentMode {
+          remove(.textSegmentOptions)
+        }
+        
+        insert(.init(opt.kind))
+      }
+      for opt in sequence.removing {
+        remove(.init(opt.kind))
+      }
```

Comment:
@hamishknight can you double check the logic and make sure this does what you expect it to do?

Reply:
Looks good to me, though I actually realized that the `x` and `xx` options are also mutually exclusive, which we'll likely want to handle here (e.g unsetting `x` unsets `xx`). Though I'm happy to take care of that when implementing the parsing for PCRE extended syntax as we'll need some way of tracking the `x` and `xx` options in the parser for that. Perhaps the parser could use this semantic model to do that, or else we maybe could implement a simpler version of it just for the `x` and `xx` options.

Reply:
`x` and `xx` are necessarily irrelevant by this stage, right? They affect parsing, but don't have any continued impact during compilation. Do we want to allow them in this grouped/intra-literal position, or should they only be permitted as "global" options?

Reply:
Yeah they're only relevant to the parser, my thinking was that if we wanted the parser to also be able to use this semantic model to track which of the syntax options is enabled in a given scope, then we could handle them here, or if the parser implements its own logic for that, we could disregard them here and just have this be specific to the matching engine.

> Do we want to allow them in this grouped/intra-literal position, or should they only be permitted as "global" options?

As I understand it, they can be set and unset arbitrarily throughout the regex, e.g `(?x) a b c (?-x)d e f`, only the whitespace in ` a b c ` is treated as non-semantic

Reply:
Thanks!

> can implement whitespace discarding in a follow-up

Shouldn't this be a case of the parser producing `Trivia` nodes, and the matching engine disregarding them? FWIW I have a patch that implements that https://github.com/apple/swift-experimental-string-processing/commit/a83ca4d6671127402f6d162c94acfc7589a5821e. I'm not sure it could be done completely in the matching engine, as PCRE has some funky rules that means e.g `x * +` is a quantification under extended syntax (my patch doesn't yet implement it tho).